x-pack/legacy/plugins/integrations_manager/server/registry.ts

```diff
@@ -67,42 +73,48 @@ async function fetchArchiveBuffer(key: string): Promise<Buffer> {
   return getResponseStream(`${REGISTRY}/package/${key}`).then(streamToBuffer);
 }
 
-export async function getObjects(pkgkey: string, desiredType?: string): Promise<SavedObject[]> {
-  const paths = await getArchiveInfo(`${pkgkey}.tar.gz`);
-  const toBeSavedObjects = paths.reduce((map, path) => {
-    collectReferences(map, { path, desiredType });
-    return map;
-  }, new Map());
+export async function getObjects(
```

Comment:
Condensed `getObjects` & `collectReferences` into one function. Added a ton of comments. 

Went with nested `for`s vs several `map`, `reduce` calls and `flatten`. For context, these might be the first `for` loops I’ve written in 5 years. I’m not overly concerned with performance but would like to skip the garbage creation/collection of traversing & transforming as much as possible. I also find this the most clear way to express what we’re doing.

Reply:
Love these comments. Totally fine with the for loops, I don't love mutating objects that get passed to other functions, so having this all in one function really helps follow that mutation thread for me.

Reply:
This feels like a function I'd love to have tests for. Can we log a ticket to do that soon, if not now? Not concerned about 100% coverage or anything but testing a few of these functions would be a great start.

Reply:
For sure. Created [this very minimal issue](https://github.com/elastic/kibana/issues/40802) but feel free to add any others you think of.